### ISSUE TITLE
How to register custom `DataSet`s for scoped/local seed propagation given by `UseSeed`

### ISSUE THREAD

**Issue by ChristoWolf** [bchavez/Bogus#405]
### Version Information
| Software                       | Version(s) |
| ------------------------| ---------- |
| Bogus NuGet Package  | 33.1.1 |
| .NET SDK                 | 5.0.403 |
| Windows OS              | Version 20H2 |
| Visual Studio?              | Microsoft Visual Studio Enterprise 2022 (64-bit) Version 17.0.1 |

### What locale are you using with Bogus?
`en`

### What's the problem?
Hi!
I am not really sure how to make a custom `DataSet` use a `Faker<T>` instance's configured scoped/local seed (configured via its `UseSeed(int)` option, added during [this PR](https://github.com/bchavez/Bogus/pull/101) due to [this issue](https://github.com/bchavez/Bogus/issues/100)).
If I understand correctly, this is because its `FakerHub` (of course; or only if I am missing something) cannot (seed-) notify the custom `DataSet`.
E.g. if we have a `CustomDataSet` (similarly to how it is suggested [in this example](https://github.com/bchavez/Bogus/blob/master/Examples/ExtendingBogus/FoodDataSet.cs)), then using the `this.Random` for generating does not give seeded values, because of [the randomizer never being set](https://github.com/bchavez/Bogus/blob/b093964bb895e11d051ad53eba13275ffb103e53/Source/Bogus/Faker.cs#L55).

### What possible solutions have you considered?
As I understand it, it should be possible for `Faker` derivates, because `this.Notifier` would be accessible, thus one could add
```csharp
this.CustomDataSet = this.Notifier.Flow(new CustomDataSet());
```
in there.
But this does not work for `Faker<T>`, because the `Notifier` of its `FakerHub` is not accessible.

I have also considered explicitly passing the `Randomizer` instance to the custom dataset, but that would disable the notifier mechanism there, if I understand correctly.

### Do you have sample code to show what you're trying to do?
Not sure how to do this in a short amount of time in LinqPad.
But simply using the [`Food` dataset](https://github.com/bchavez/Bogus/blob/master/Examples/ExtendingBogus/FoodDataSet.cs) from a `Faker<T>` instance with configured constant local seed yields different results for each test/app execution because the randomizer is never explicitly [set, thus a new `Randomizer` instance without the seed] is created and attached to it.(https://github.com/bchavez/Bogus/blob/b093964bb895e11d051ad53eba13275ffb103e53/Source/Bogus/Faker.cs#L55)

**Comment by ChristoWolf** [comment]
I just traced back my thoughts, and maybe it is possible similarly to
```csharp
public Internet(string locale = "en") : base(locale)
{
    this.Name = this.Notifier.Flow(new Name(locale));
    ....
}
```
if I understand correctly?

I will try this out.

**Comment by bchavez** [comment]
Hi @ChristoWolf, thank you for creating an issue.

* Do you have a code sample that demonstrates what you are trying to achieve?

I have a hunch about what you are trying to do, but I can't be totally sure from the description above ... could you please post some sample code? I want to make sure I'm helping with the right answer to your question.

Thank you,
Brian

**Comment by bchavez** [comment]
@ChristoWolf, I think this is what you're looking for; a way to reset the seed deterministically at the `Faker<T>` level with a custom, data set. The basic idea is to execute `ContextHelper.GetOrSet(faker, () => new Food())` as early as possible instead of at `.RuleFor(__, f => f.Food()... )` rule invocation time.

The `Faker<User>.AttachFood()` method encapsulates this earlier invocation before `.RuleFor(__, f => f.Food()... )` and the subsequent `.UseSeed()` flows the local seed randomizer to `new Food()` before `.RuleFor` invocations begin.

```csharp
void Main()
{
   var f = new Faker<User>()
            .AttachFood()
            .UseSeed(1111)
            .RuleFor(x => x.Name, f => f.Name.FirstName())
            .RuleFor(x => x.Candy, f => f.Food().Candy());
            
            
   f.Generate().Dump();
   
   f.UseSeed(1111);
   
   f.Generate().Dump();
}

public class User
{
   public string Name;
   public string Candy;
}

public static class ExtensionsForFood
{
   public static Faker<T> AttachFood<T>(this Faker<T> fakerT) where T : class
   {
      var fakerInternal = fakerT as IFakerTInternal;
      var faker = fakerInternal.FakerHub;
      
      ContextHelper.GetOrSet(faker, () => new Food());
      
      return fakerT;
   }
   
   public static Food Food(this Faker faker)
   {
      return ContextHelper.GetOrSet(faker, () => new Food());
   }
}

public class Food : Bogus.DataSet
{
   private static readonly string[] Candies =
      {
            "Hard candy", "Taffy", "Chocolate bar", "Stick candy",
            "Jelly bean", "Mint", "Cotton candy", "Lollipop"
         };

   public string Candy()
   {
      return this.Random.ArrayElement(Candies);
   }

   private static readonly string[] Drinks = { "Soda", "Water", "Beer", "Wine", "Coffee", "Lemonade", "Milk" };
   public string Drink()
   {
      return this.Random.ArrayElement(Drinks);
   }
}
```

![image](https://user-images.githubusercontent.com/478118/149067978-34c595ab-6ca7-4045-919c-d9a902ef5961.png)

Let me know if that helps.

**Comment by ChristoWolf** [comment]
Hi @bchavez!

First of all, thanks a ton for the incredibly fast feedback!

The solution that you provided is exactly what I was looking for, thanks!
Now in hindsight it seems so obvious, this is great.

Would it make sense for you if this example was added to the [`Food` dataset example](https://github.com/bchavez/Bogus/blob/master/Examples/ExtendingBogus/FoodDataSet.cs) and/or the README?

Thanks again!

**Comment by bchavez** [comment]
Yeah, I think that makes sense. We should probably add the extra code to the `Food` dataset example. I'll leave this ticket open until we get the work completed. Thanks again @ChristoWolf for writing up the issue.

**Comment by ChristoWolf** [comment]
If you need help with that, I can also add this to the example, just let me know.

Thanks again for the great help and of course all the stuff that you are providing with Bogus, @bchavez!

**Comment by ChristoWolf** [comment]
@bchavez:
BTW, what I ended up doing on my side was to provide a generic extension method (generic w.r.t. `T` and the custom `DataSet`), so that I can simply call it always for any custom `DataSet` if needed.

Would it make sense to add such a (non-extension) method
```csharp
public static AttachCustomDataSet<TDataSet>() where TDataSet : DataSet, new()
   {
      // Generic implementation
   }
```
to `Faker<T>` as a new feature?

**Comment by bchavez** [comment]
Hi @ChristoWolf, that's a great way to generalize the problem.

Maybe, we could add it, but I would need to really think about it. Generally, I don't change the public APIs of `Faker` and `Faker<T>` very much but when a proposal like this is made, it does take some time for me to think about the situation.

I like to make sure our public APIs that 21M downloads see are generally consistent and well-thought-out before making any changes.

Usually, my thought process is:
* Does it make sense for what we are trying to do? Or is it a hack for a deeper problem?
* How many people are affected by an issue?
* Is there any way it could be done better to make it easier on the user?
* Is it easy for someone to use the API wrong? - And if they do, what would they expect to happen that isn't happening?

I just have to really think about it before moving on to a public API change. But I do like your generic solution for resolving the original issue you had :) :sunglasses:

**Comment by ChristoWolf** [comment]
Great feature proposal thought process, makes a lot of sense of course!

Thanks again, and feel free to let me know if I can help, @bchavez!

**Comment by garcipat** [comment]
I just created my own dataset and discovered too that the seed is not respected in the dataset, even with that Attachfood approach provided.
```csharp
public class ColorDataSet : DataSet
{
    public string Hex(bool includeAlpha = false)
    {
        var r = Random.Int(0, 255);
        var g = Random.Int(0, 255);
        var b = Random.Int(0, 255);

        var color = $"#{r:X2}{g:X2}{b:X2}";

        if (includeAlpha)
        {
            var a = Random.Int(0, 255);
            return color + r.ToString("X2");
        }

        return color;
    }
}

public static ColorDataSet Color(this Faker faker)
{
   return ContextHelper.GetOrSet(faker, () => new ColorDataSet());
}

public static Faker<T> AttachColor<T>(this Faker<T> faker) where T : class
{
   var fakerInternal = faker as IFakerTInternal;
   var hub = fakerInternal.FakerHub;

   ContextHelper.GetOrSet(hub, () => new ColorDataSet());

   return faker;
}
```

This results in teh hex color being generated with different results even when the Faker<T> is created with UseSeed.
Did I miss something?

**Comment by ChristoWolf** [comment]
@garcipat: Can you please show your code where you actually attach the data set?

**Comment by ChristoWolf** [comment]
Yep, the order is important, happy that it works for you now!

**Comment by garcipat** [comment]
I'm very confused. I'm trying it again with another DataSet and I can't get it to work again. With Seeding before or after, It generates different values in my test:
```csharp
using Bogus;
using Snapshooter.Xunit;

public class DataSetExtensionsTests
{
    [Fact]
    public void AttachDataSet_ShouldGenerateTheSameObject()
    {
        var faker = new Faker<TestClass>()
            .AttachDataSet(typeof(TestDataSet))
            .UseSeed(12345678)
            .RuleFor(x => x.IntProp, f => f.Random.Int())
            .RuleFor(x => x.StringProp, f => f.TestData().TestValue());

        var generated = faker.Generate();

        generated.Should().MatchSnapshot();
    }

    private record TestClass
    {
        public int IntProp { get; set; }
        public string StringProp { get; set; }
    }

    internal class TestDataSet : DataSet
    {
        private readonly string[] _testValues
            = Enumerable.Range(1, 100)
                .Select(i => "T" + i)
                .ToArray();

        public string TestValue()
        {
            return Random.ArrayElement(_testValues);
        }
    }
}

internal static class FakerExtensions
{
    public static TestDataSet TestData(this Faker faker)
    {
        return ContextHelper.GetOrSet(faker, () => new TestDataSet());
    }
}
```

**Comment by ChristoWolf** [comment]
Hi @BenjaBobs!

This is done via so called notifier flows, which propagate seed information to dependency datasets.
See e.g. [how the `Address` dataset depends on the `Name` dataset and how it configures the notifier flow](https://github.com/bchavez/Bogus/blob/905a7e753ed2d7bfd42fc5099a50278adf45b467/Source/Bogus/DataSets/Address.cs#L21).

**Comment by BenjaBobs** [comment]
Yes, and as far as I can tell, this allows the root `Faker` to notify/propagate new values for `Randomizer`, however in my case I'm not changing the `Randomizer`, I'm changing the `.Date.LocalSystemClock`.
I suppose what I'm looking for is a way to tell an instance of a `Faker<>` to use a specific `Faker` as it's "_base_".

Maybe this isn't the way I'm supposed to set up such a chain of generation?

**Comment by garcipat** [comment]
> Yes, and as far as I can tell, this allows the root `Faker` to notify/propagate new values for `Randomizer`, however in my case I'm not changing the `Randomizer`, I'm changing the `.Date.LocalSystemClock`. I suppose what I'm looking for is a way to tell an instance of a `Faker<>` to use a specific `Faker` as it's "_base_".
> 
> Maybe this isn't the way I'm supposed to set up such a chain of generation?

What you could do is to define a factory fro the Faker that takes in your FakeTimeProvider and every instance of the Faker you create. gets the currenlty set Time as Reference DateTime. I'm doing that as well to create multiple fakers with one seed, but to ensure they have not duplicate seeds:

```csharp
public class FakerFactory
{
    private readonly Random _seeder;
    private readonly List<DataSet> _dataSets = new();
    private readonly TimeProvider _timeProvider;

    public FakerFactory(int seed, TimeProvider timeProvider) // Insert your provider here
    {
        _seeder = new Random(seed);
        _timeProvider = timeProvider;
    }

    public Faker<T> Create<T>(Action<Faker<T>> withRules) where T : class
    {
        var faker = new Faker<T>();
        var method = typeof(FakerExtensions).GetMethod(nameof(FakerExtensions.AttachDataSet))!;

        // Attach has to be before seeding
        foreach (var dataSets in _dataSets)
        {
            var generic = method.MakeGenericMethod(typeof(T), dataSets.GetType());
            faker = (Faker<T>)generic.Invoke(method, new object[] { faker })!;
        }

        faker = faker
            .UseSeed(_seeder.Next())
            .UseDateTimeReference(_timeProvider.GetUtcNow().DateTime); // Get Date reference for this Faker instance here

        withRules(faker);

        faker.AssertConfigurationIsValid();
        return faker;
    }

    public FakerFactory WithDataSet<TDataSet>()
        where TDataSet : DataSet, new()
    {
        _dataSets.Add(new TDataSet());
        return this;
    }
}
```

(This is also including how I attach datasets now)
